File: src/main/java/jv/collection/stacks/StackArray.java
package jv.collection.stacks;

import java.util.Iterator;

public class StackArray<T> implements Stack<T> {
    private int n = 0;
    private final T[] data;

    @SuppressWarnings("unchecked")
    public StackArray(int capacity) {
        data = (T[]) new Object[capacity];
    }

    @Override
    public void push(T item) {
        data[n++] = item;
    }

    @Override
    public T pop() {
        T elem = data[--n];
        data[n] = null;
        return elem;
    }

    @Override
    public boolean isEmpty() {
        return n == 0;
    }

    @Override
    public Iterator<T> iterator() {
        return new ArrayIterator<>(data, n);
    }
}

File: src/main/scala/sc/collection/stacks/StackArray.scala
package sc.collection.stacks

class StackArray[T >: Null](capacity: Int)(implicit m: Manifest[T]) extends Stack[T] {
  private final val data = new Array[T](capacity)
  private var n = 0

  def push(item: T) {
    data(n) = item
    n += 1
  }

  def pop(): T = {
    n -= 1
    val elem = data(n)
    data(n) = null
    elem
  }

  def isEmpty: Boolean = n == 0

}